File: python/src/azoth/eos/reference/srk_kappa.py
"""``eos.srk_kappa`` - the Soave-Redlich-Kwong attraction-parameter coefficient.

```text
kappa = 0.48 + 1.574*omega - 0.176*omega**2
```

Spec: ``specs/calcs/eos/srk_kappa.toml``, which carries the citation and why a negative
``kappa`` is returned with a warning rather than refused.

Soave's coefficient is the same *form* as Peng-Robinson's with three different
constants, which is why the alpha functions share ``azoth.eos.alpha_term``.
"""

from __future__ import annotations

from azoth._registry_gen import spec as _spec_for
from azoth.core.range import apply_checks, checks_for
from azoth.core.result import SrkKappaResult
from azoth.core.warnings import Warning

CALC_ID = "eos.srk_kappa"


def srk_kappa(omega: float) -> SrkKappaResult:
    """The Soave-Redlich-Kwong alpha-function coefficient for a pure component.

    Args:
        omega: the Pitzer acentric factor. Dimensionless, and supplied by the caller.

    Returns:
        The coefficient, and any caveats. A negative ``kappa`` comes back carrying
        ``OUT_OF_VALID_RANGE`` rather than as an error.

    Example:
        >>> r = srk_kappa(0.152)
        >>> round(r.kappa, 9)
        0.715181696
        >>> r.is_clean
        True
    """
    spec = _spec_for(CALC_ID)
    checks = checks_for(spec)
    warnings: list[Warning] = []

    values = {"omega": omega}

    apply_checks(checks.on_input, values.get, warnings)

    kappa = 0.48 + 1.574 * omega - 0.176 * omega * omega

    apply_checks(checks.derived, lambda name: kappa if name == "kappa" else None, warnings)

    return SrkKappaResult(kappa=kappa, warnings=tuple(warnings))

File: crates/azoth-eos/src/srk_kappa.rs
//! `eos.srk_kappa` - the Soave-Redlich-Kwong attraction-parameter coefficient.
//!
//! ```text
//! kappa = 0.48 + 1.574*omega - 0.176*omega**2
//! ```
//!
//! Spec: `specs/calcs/eos/srk_kappa.toml`, which carries the citation and why a
//! negative `kappa` is returned with a warning rather than refused.
//!
//! Soave's coefficient is the whole temperature dependence of the SRK attraction term,
//! in one number, and it is the same *form* as Peng-Robinson's with three different
//! constants - which is why the alpha functions themselves share [`crate::alpha_term`].

use azoth_core::{Result, apply_checks};

use crate::results::SrkKappaResult;
use crate::spec_gen;

/// The Soave-Redlich-Kwong alpha-function coefficient for a pure component.
///
/// `omega` is the Pitzer acentric factor, supplied by the caller. The returned
/// [`SrkKappaResult`] carries warnings but no failure path beyond a malformed spec: the
/// polynomial is defined for every real `omega`.
///
/// # Example
/// ```
/// use azoth_eos::srk_kappa;
///
/// // Propane-like.
/// let r = srk_kappa(0.152)?;
/// assert!((r.kappa - 0.715181696).abs() < 1e-12);
/// assert!(r.warnings.is_empty());
/// # Ok::<(), azoth_core::AzothError>(())
/// ```
pub fn srk_kappa(omega: f64) -> Result<SrkKappaResult> {
    let spec = &spec_gen::SRK_KAPPA_SPEC;
    let mut warnings = Vec::new();

    apply_checks(
        spec.input_checks(),
        |quantity| match quantity {
            "omega" => Some(omega),
            _ => None,
        },
        &mut warnings,
    )?;

    let kappa = 0.48 + 1.574 * omega - 0.176 * omega * omega;

    apply_checks(
        spec.derived_checks(),
        |name| (name == "kappa").then_some(kappa),
        &mut warnings,
    )?;

    Ok(SrkKappaResult { kappa, warnings })
}
